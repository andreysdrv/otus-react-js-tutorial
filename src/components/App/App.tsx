import React, { ChangeEvent, Component, FormEvent } from "react";
import "./App.css";
import { Main } from "../../Layouts/Main";
import { List } from "../List";
import { ProductItem } from "../ProductItem";
import MOCK_PRODUCTS_LIST from "../../utils/_ProductsList.json";
import { IProduct, IProducts } from "../../types/types";
import { AddForm } from "../AddForm";

export default class App extends Component<
  Record<string, never>,
  { products: IProduct[]; inputVal: string }
> {
  constructor(props: any) {
    super(props);
    this.state = {
      products: [],
      inputVal: "",
    };

    this.getProducts = this.getProducts.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentDidMount() {
    this.getProducts().then((response: IProducts) => {
      this.setState({ products: response.products });
    });

    document.querySelector("form")?.addEventListener("submit", () => {
      alert("success");
    });
  }

  componentDidUpdate(
    prevProps: Readonly<Record<string, never>>,
    prevState: Readonly<{ products: IProduct[]; inputVal: string }>
  ) {
    if (prevState.inputVal === "secret") {
      this.setState({
        products: [
          { id: Math.random(), completed: false, title: "secret" },
          { id: Math.random(), completed: false, title: "secret" },
          { id: Math.random(), completed: false, title: "secret" },
        ],
      });
    }
  }

  componentWillUnmount() {
    return () => {
      document.querySelector("form")?.removeEventListener("submit", () => {
        alert("success");
      });
    };
  }

  getProducts(): Promise<IProducts> {
    return new Promise<IProducts>((resolve) => {
      setTimeout(() => {
        resolve(MOCK_PRODUCTS_LIST);
      });
    });
  }

  handleChange(e: ChangeEvent<HTMLInputElement>) {
    this.setState({ inputVal: e.target.value });
  }

  handleSubmit(e: FormEvent<HTMLFormElement>) {
    e.preventDefault();

    this.setState((prev) => ({
      products: [
        { id: Math.random(), completed: false, title: this.state.inputVal },
        ...prev.products,
      ],
      inputVal: "",
    }));
  }

  render() {
    return (
      <Main>
        <AddForm
          value={this.state.inputVal}
          onChange={this.handleChange}
          onSubmit={this.handleSubmit}
        />
        <List
          className={"products-list"}
          items={this.state.products}
          renderItem={(product: IProduct) => (
            <ProductItem product={product} key={product.id} />
          )}
        />
      </Main>
    );
  }
}
