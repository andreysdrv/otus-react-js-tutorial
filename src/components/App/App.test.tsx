import React from "react";
import { render } from "@testing-library/react";
import App from "./App";

describe("App snapshot tests", () => {
  it("base snapshot", () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  it("some data snapshot", () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });
});
