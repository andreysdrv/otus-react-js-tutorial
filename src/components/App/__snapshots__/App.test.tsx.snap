// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot tests base snapshot 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="field"
      style="width: 300px; height: 300px;"
    >
      <div
        class="cell"
      >
        1
      </div>
      <div
        class="cell"
      >
        2
      </div>
      <div
        class="cell"
      >
        3
      </div>
      <div
        class="cell"
      >
        4
      </div>
      <div
        class="cell"
      >
        5
      </div>
      <div
        class="cell"
      >
        6
      </div>
      <div
        class="cell"
      >
        7
      </div>
      <div
        class="cell"
      >
        8
      </div>
      <div
        class="cell"
      >
        9
      </div>
    </div>
  </div>
</div>
`;

exports[`App snapshot tests some data snapshot 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="field"
      style="width: 300px; height: 300px;"
    >
      <div
        class="cell"
      >
        1
      </div>
      <div
        class="cell"
      >
        2
      </div>
      <div
        class="cell"
      >
        3
      </div>
      <div
        class="cell"
      >
        4
      </div>
      <div
        class="cell"
      >
        5
      </div>
      <div
        class="cell"
      >
        6
      </div>
      <div
        class="cell"
      >
        7
      </div>
      <div
        class="cell"
      >
        8
      </div>
      <div
        class="cell"
      >
        9
      </div>
    </div>
  </div>
</div>
`;
